README.md
SSH signatures for Erlang
=====

Implementation of [SSH signatures][ssh-keygen-sign] in Erlang. It uses only
stuff distributed with OTP, so no external dependencies needed.

[ssh-keygen-sign]: https://man.openbsd.org/ssh-keygen#Y~4

Currently supported algorithms:

- RSA
- Ed25519
- Ed448 - not tested, as my implementation of OpenSSH do not support Ed448 keys

## Usage

There are just 3 exported functions:

- `sign/{3,4}` which allows signing data
- `verify/2` that verifies the signature for given data and outputs details
  about signature

```erlang
% First we need the private key that we will use for signing.
% For the purpose of this example just use RSA-4096
SecretKey = public_key:generate_key({rsa, 4096, 3}),

Data = <<"Foo">>,

% Sign data using our key. 3rd argument there is a namespace, that must be
% non-empty string.
Signature = ssh_signature:sign(Data, SecretKey, "text"),

% The created signature is already in armoured (ASCII-only) format.

% Now we can check if the signature is correct
{ok, #{public_key := PubKey, ns := <<"test">>, signature := Sig}} =
    ssh_signature:verify(Data, Signature).
% Notice that we do not pass public key to verify/2, it is left to the user to
% check whether the returned public key is trusted.
```

## License

See [LICENSE](LICENSE)


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="v0.1.0"></a>
## v0.1.0 - 2022-08-20
### Bug Fixes
- conditionaly use removed function ([`784871a`](https://github.com/hauleth/ssh_signature/commit/784871a584db40a6a2e954b5d6093babecf0a28e))
- install OpenSSL GitHub Actions ([`09dd972`](https://github.com/hauleth/ssh_signature/commit/09dd9727b849c5cb90f411041f377a80839a0d40))

### Documentation
- add protocol description to the docs ([`efd03d2`](https://github.com/hauleth/ssh_signature/commit/efd03d279772c9ff159748cedb33c9c780e30ba5))
- add example in README ([`b694f60`](https://github.com/hauleth/ssh_signature/commit/b694f6004b7e64a0ded835a59098e5e9e8b3c04b))


[Unreleased]: https://github.com/hauleth/ssh_signature/compare/v0.1.0...HEAD


test/ssh_signature_SUITE.erl
-module(ssh_signature_SUITE).

-include_lib("public_key/include/public_key.hrl").

% -compile(export_all).

-export([
    all/0,
    groups/0,
    init_per_group/2,
    end_per_group/2,
    can_verify_signed/1,
    signature_can_be_verified_by_openssh/1,
    openssh_can_be_verified/1
]).

-include_lib("stdlib/include/assert.hrl").
-include_lib("common_test/include/ct.hrl").

all() ->
    [
        {group, ed25519},
        {group, rsa2048},
        {group, rsa3072},
        {group, rsa4096}
    ].

groups() ->
    Tests = [
        can_verify_signed,
        signature_can_be_verified_by_openssh,
        openssh_can_be_verified
    ],
    Hashes = [sha256, sha512],
    HashTests = [{Hash, [parallel], Tests} || Hash <- Hashes],
    Algos = [ed25519, rsa2048, rsa3072, rsa4096],
    [{Algo, [parallel], HashTests} || Algo <- Algos].

init_per_group(basic, Config) ->
    Config;
init_per_group(GroupName, Config) when
    GroupName =:= sha256; GroupName =:= sha512
->
    [{hash, GroupName} | Config];
init_per_group(GroupName, Config) when
    GroupName =:= ed25519;
    GroupName =:= rsa2048;
    GroupName =:= rsa3072;
    GroupName =:= rsa4096
->
    try
        Key = create_key(GroupName),
        ct:log("Key = ~p.", [Key]),
        [{key, Key}, {algo, GroupName} | Config]
    catch
        _C:_R:_S ->
            {skip, "Algorithm not supported"}
    end.

create_key(ed25519) ->
    public_key:generate_key({namedCurve, ?'id-Ed25519'});
create_key(ed448) ->
    public_key:generate_key({namedCurve, ?'id-Ed448'});
create_key(rsa2048) ->
    public_key:generate_key({rsa, 2048, 3});
create_key(rsa3072) ->
    public_key:generate_key({rsa, 3072, 3});
create_key(rsa4096) ->
    public_key:generate_key({rsa, 4096, 3}).

end_per_group(_, Config) ->
    Config.

can_verify_signed(Config) ->
    Key = ?config(key, Config),
    Hash = ?config(hash, Config),
    Data = crypto:strong_rand_bytes(256),

    PubKey = ssh_signature:priv_to_public(Key),

    Signature = ssh_signature:sign(Data, Key, <<"file">>, #{hash => Hash}),
    ct:log("Signature = ~s.", [Signature]),

    ?assertMatch(
        {ok, #{public_key := PubKey}}, ssh_signature:verify(Data, Signature)
    ).

signature_can_be_verified_by_openssh(Config) ->
    ct:make_priv_dir(),
    Key = ?config(key, Config),
    Hash = ?config(hash, Config),
    PrivDir = ?config(priv_dir, Config),

    DataFile = filename:join(PrivDir, "data"),
    AllowedSignersFile = filename:join(PrivDir, "key"),
    SigFile = filename:join(PrivDir, "data.sig"),

    Data = crypto:strong_rand_bytes(256),
    ExportedKey = encode([{ssh_signature:priv_to_public(Key), []}]),

    Signature = ssh_signature:sign(Data, Key, <<"file">>, #{hash => Hash}),

    ct:log("DataB64 = ~s.", [base64:encode(Data)]),
    ct:log("ExportedKey = ~s.", [ExportedKey]),
    ct:log("Signature = ~s.", [Signature]),
    file:write_file(DataFile, Data),
    file:write_file(AllowedSignersFile, ["test@example.com ", ExportedKey]),
    file:write_file(SigFile, Signature),

    ?assert(
        ssh_keygen(
            [
                "-Y",
                "verify",
                "-n",
                "file",
                "-f",
                AllowedSignersFile,
                "-I",
                "test@example.com",
                "-s",
                SigFile
            ],
            Data
        )
    ),

    Config.

openssh_can_be_verified(Config) ->
    ct:make_priv_dir(),
    Algo = ?config(algo, Config),
    Hash = ?config(hash, Config),
    PrivDir = ?config(priv_dir, Config),

    PrivKeyFile = filename:join(PrivDir, "key"),
    PubKeyFile = filename:join(PrivDir, "key.pub"),
    DataFile = filename:join(PrivDir, "data"),
    SigFile = filename:join(PrivDir, "data.sig"),

    {Type, ExtraArgs} = openssh_keygen_args(Algo, Hash),
    true = ssh_keygen(
        ["-t", Type, "-f", PrivKeyFile, "-N", ""] ++ ExtraArgs, ""
    ),

    {ok, PubKeyPEM} = file:read_file(PubKeyFile),

    ct:log("PubKeyPEM = ~s.", [PubKeyPEM]),

    Data = crypto:strong_rand_bytes(256),
    ct:log("Data = ~s.", [base64:encode(Data)]),
    file:write_file(DataFile, Data),

    [{PK, _}] = decode(PubKeyPEM),

    true = ssh_keygen(
        ["-Y", "sign", "-f", PrivKeyFile, "-n", "test", DataFile], ""
    ),

    {ok, Signature} = file:read_file(SigFile),

    {ok, Result} = ssh_signature:verify(Data, Signature),
    ?assertMatch(#{public_key := PK}, Result),

    Config.

openssh_keygen_args(ed25519, _) -> {"ed25519", []};
openssh_keygen_args(rsa2048, H) -> {rsa(H), ["-b", "2048"]};
openssh_keygen_args(rsa3072, H) -> {rsa(H), ["-b", "3072"]};
openssh_keygen_args(rsa4096, H) -> {rsa(H), ["-b", "4096"]}.

rsa(sha256) -> "rsa-sha2-256";
rsa(sha512) -> "rsa-sha2-512".

ssh_keygen(Args, Input) ->
    Stdout = fun(_, _, Data) ->
        ct:pal("ssh-keygen -> ~s", [Data])
    end,
    exec:start(),
    Exec = os:find_executable("ssh-keygen"),
    {ok, Pid, _OsPid} = exec:run([Exec | Args], [
        stdin, monitor, {stdout, Stdout}
    ]),
    case string:is_empty(Input) of
        false ->
            exec:send(Pid, Input),
            exec:send(Pid, eof);
        true ->
            ok
    end,
    receive
        {'DOWN', _, process, _, normal} ->
            true
    after 2000 ->
        false
    end.

-if(?OTP_RELEASE < 24).
encode(Key) ->
    public_key:ssh_encode(Key, openssh_public_key).

decode(Key) ->
    public_key:ssh_decode(Key, openssh_public_key).
-else.
encode(Key) ->
    ssh_file:encode(Key, openssh_key).

decode(Key) ->
    ssh_file:decode(Key, openssh_key).
-endif.


src/ssh_signature.erl
-module(ssh_signature).

-dialyzer([
    {no_improper_lists, split/1},
    % Needed due to fact that OTP 25 removed `{ed_pri, _, _, _}' key format
    {no_match, [
        ed_to_pub/1,
        pk_sign/3,
        pk_verify/4,
        priv_to_public/1,
        sig_type/2,
        type_sig/3
    ]},
    {nowarn_function, [pk_sign/3, pk_verify/4]}
]).

-include_lib("public_key/include/public_key.hrl").

-export([sign/3, sign/4]).
-export([verify/2]).

-ifdef(TEST).
-export([priv_to_public/1, decode/1]).
-endif.

-export_type([namespace/0, hash_algorithm/0]).

-define(MAGIC_PREAMBLE, "SSHSIG").
-define(SIG_VERSION, 16#01).
-define(UINT32(X), (X):32 / unsigned - big - integer).
-define(STRING(X), ?UINT32(size(X)), (X) / binary).
-define(BEGIN, "-----BEGIN SSH SIGNATURE-----").
-define(END, "-----END SSH SIGNATURE-----").

-type namespace() :: unicode:chardata().
-type hash_algorithm() :: sha256 | sha512.

%% @equiv sign(Data, Key, NS, #{})
sign(Data, Key, NS) -> sign(Data, Key, NS, #{}).

%% @doc Sign `Data' using SSH signature format with `Key'.
%%
%% The `NS' must be not empty.
%%
%% == Options ==
%%
%% <ul>
%%      <li>`hash' - hash algorithm used on input data. Can be either `sha256'
%%      or `sha512'. Defaults to `sha512'.</li>
%% </ul>
%% @end
-spec sign(iodata(), public_key:private_key(), namespace(), Opts) ->
    unicode:chardata()
when
    Opts :: #{
        hash => hash_algorithm()
    }.
sign(Data, Key, NS, Opts) ->
    NS0 = iolist_to_binary(NS),
    case NS0 of
        <<>> -> error({badarg, empty_namespace});
        _ -> ok
    end,
    Algo = maps:get(hash, Opts, sha512),
    Reserved = <<"">>,
    Body = body(Data, NS0, Reserved, Algo),
    Signature = pk_sign(Body, Algo, Key),
    SigType = sig_type(Key, Algo),
    Sig = <<?STRING(SigType), ?STRING(Signature)>>,
    EncPub = encode(priv_to_public(Key)),
    Result =
        <<?MAGIC_PREAMBLE, ?UINT32(?SIG_VERSION), ?STRING(EncPub), ?STRING(NS0),
            ?STRING(Reserved), ?STRING(atom_to_binary(Algo, utf8)),
            ?STRING(Sig)>>,
    iolist_to_binary([?BEGIN, $\n, split(base64:encode(Result)), $\n, ?END]).

%% @doc Verify `Signature' of `Data'.
%%
%% Notice that this function do not check authenticity of the provided key. That
%% is left to the user to check whether key used for signing match the
%% requirements.
%%
%% @end
-spec verify(iodata(), unicode:chardata()) -> {ok, Result} | {error, term()} when
    Result :: #{
        ns => namespace(),
        public_key => public_key:public_key(),
        signature => binary()
    }.
verify(Data, Signature) ->
    Sig0 = string:trim(Signature),
    Sig1 = iolist_to_binary(Sig0),
    Size = byte_size(Sig1) - length(?BEGIN) - length(?END) - 2,
    case Sig1 of
        <<?BEGIN, $\n, Sig2:Size/binary, $\n, ?END>> ->
            Sig3 = base64:decode(Sig2),
            case parse(Sig3) of
                {ok, #{
                    ns := NS,
                    pk := PublicKey,
                    reserved := R,
                    signature := {_, Sig},
                    hash_algorithm := Algo
                }} ->
                    Body = body(Data, NS, R, Algo),
                    case pk_verify(Body, Algo, Sig, PublicKey) of
                        true ->
                            {ok, #{
                                ns => NS,
                                public_key => PublicKey,
                                signature => Sig
                            }};
                        false ->
                            {error, invalid_signature}
                    end;
                {error, _} = Error ->
                    Error
            end;
        _ ->
            {error, invalid_armour}
    end.

parse(
    <<?MAGIC_PREAMBLE, ?UINT32(Version), ?UINT32(EPS), EncPub:EPS/binary,
        ?UINT32(NSS), NS:NSS/binary, ?UINT32(RS), R:RS/binary, ?UINT32(SAlgoS),
        SAlgo:SAlgoS/binary, ?UINT32(SigDS), ?UINT32(SAS), SigAlgo:SAS/binary,
        ?UINT32(SigS), Sig:SigS/binary>>
) when
    SigDS =:= SAS + 4 + SigS + 4,
    (SAlgo =:= <<"sha256">> orelse SAlgo =:= <<"sha512">>)
->
    PubKey = decode(EncPub),
    Algo =
        case SAlgo of
            <<"sha256">> -> sha256;
            <<"sha512">> -> sha512
        end,
    case type_sig(SigAlgo, PubKey, Algo) of
        true ->
            {ok, #{
                version => Version,
                pk => PubKey,
                ns => NS,
                reserved => R,
                hash_algorithm => Algo,
                signature => {SigAlgo, Sig}
            }};
        _ ->
            {error, type_mismatch}
    end;
parse(_) ->
    {error, invalid_format}.

pk_sign(Body, _, {ed_pri, _, _, _} = Key) ->
    public_key:sign(Body, none, Key);
pk_sign(Body, _, #'ECPrivateKey'{} = Key) ->
    public_key:sign(Body, none, Key);
pk_sign(Body, Algo, #'RSAPrivateKey'{} = Key) ->
    public_key:sign(Body, Algo, Key).

pk_verify(Body, _, Sig, {ed_pub, _, _} = Key) ->
    public_key:verify(Body, none, Sig, Key);
pk_verify(Body, _, Sig, {#'ECPoint'{}, _} = Key) ->
    public_key:verify(Body, none, Sig, Key);
pk_verify(Body, Algo, Sig, #'RSAPublicKey'{} = Key) ->
    public_key:verify(Body, Algo, Sig, Key).

body(Data, NS, R, Algo) ->
    H = crypto:hash(Algo, Data),
    <<?MAGIC_PREAMBLE, ?STRING(NS), ?STRING(R),
        ?STRING(atom_to_binary(Algo, utf8)), ?STRING(H)>>.

sig_type({ed_pri, Type, _Pub, _Pri}, _Algo) ->
    <<"ssh-", (atom_to_binary(Type, utf8))/binary>>;
sig_type(#'ECPrivateKey'{parameters = {namedCurve, ?'id-Ed25519'}}, _) ->
    <<"ssh-ed25519">>;
sig_type(#'ECPrivateKey'{parameters = {namedCurve, ?'id-Ed448'}}, _) ->
    <<"ssh-ed448">>;
sig_type(#'RSAPrivateKey'{}, sha256) ->
    <<"rsa-sha2-256">>;
sig_type(#'RSAPrivateKey'{}, sha512) ->
    <<"rsa-sha2-512">>.

type_sig(<<"ssh-ed25519">>, {ed_pub, ed25519, _}, _) -> true;
type_sig(<<"ssh-ed25519">>, {_, {namedCurve, ?'id-Ed25519'}}, _) -> true;
type_sig(<<"ssh-ed448">>, {_, {namedCurve, ?'id-Ed448'}}, _) -> true;
type_sig(<<"ssh-ed448">>, {ed_pub, ed448, _}, _) -> true;
type_sig(<<"rsa-sha2-256">>, #'RSAPublicKey'{}, sha256) -> true;
type_sig(<<"rsa-sha2-512">>, #'RSAPublicKey'{}, sha512) -> true;
type_sig(_, _, _) -> false.

-spec priv_to_public(public_key:private_key()) -> public_key:public_key().
priv_to_public({ed_pri, _, _, _} = Priv) ->
    ed_to_pub(Priv);
priv_to_public(#'ECPrivateKey'{} = Priv) ->
    ed_to_pub(Priv);
priv_to_public(#'RSAPrivateKey'{modulus = Mod, publicExponent = Exp}) ->
    #'RSAPublicKey'{modulus = Mod, publicExponent = Exp};
priv_to_public(Other) ->
    Other.

-if(?OTP_RELEASE < 25).
ed_to_pub({ed_pri, Type, Pub, _Priv}) ->
    {ed_pub, Type, Pub};
ed_to_pub(#'ECPrivateKey'{
    parameters = {namedCurve, ?'id-Ed25519'}, publicKey = Pub
}) ->
    {ed_pub, ed25519, Pub};
ed_to_pub(#'ECPrivateKey'{
    parameters = {namedCurve, ?'id-Ed448'}, publicKey = Pub
}) ->
    {ed_pub, ed448, Pub}.
-else.
ed_to_pub({ed_pri, Type, Pub, _Priv}) ->
    {#'ECPoint'{point = Pub}, Type};
ed_to_pub(#'ECPrivateKey'{parameters = Type, publicKey = Pub}) ->
    {#'ECPoint'{point = Pub}, Type}.
-endif.

-spec encode(public_key:public_key()) -> binary().
-spec decode(binary()) -> public_key:public_key().
-if(?OTP_RELEASE < 24).
encode(Key) ->
    public_key:ssh_encode(Key, ssh2_pubkey).

decode(Key) ->
    public_key:ssh_decode(Key, ssh2_pubkey).
-else.
encode(Key) ->
    ssh_file:encode(Key, ssh2_pubkey).

decode(Key) ->
    ssh_file:decode(Key, ssh2_pubkey).
-endif.

split(<<D:70/binary, Rest/binary>>) ->
    [D, $\n | split(Rest)];
split(Rest) ->
    Rest.
